### ISSUE TITLE
Investigate possible inappropriate access of UCSB ETDs collection

### ISSUE THREAD

**Issue by elopatin-uc3** [CDLUC3/mrt-doc#165]
Background:
From Chrissy Rissmeyer at UCSB – 
"On Sep 18, 2019, credentials for the UCSB Library ETD collection in Merritt were uploaded to a work-in-progress branch on the ucsblibrary GitHub's public alexandria project. The credentials were provided by to UCSB for use in development on an ongoing Electronic Theses and Dissertations Ingest project. These credentials provide access to otherwise private data in Merritt. This includes personally-identifying information for UCSB students available in ProQuest metadata (names, addresses, email, and citizenship), and embargoed or otherwise unpublished theses and dissertations. Embargoed documents sometimes contain sensitive information and/or material under copyright."

Summary:
Per UCSB's request, we need to investigate and attempt to determine if any inappropriate access to embargoed or unpublished ETDs was had. Initial thoughts are we should be looking for patterns of access to the collection that indicate scraping of any sort. 

- [x] Obtain a list of ARKs for the UCSB ETDs collection, and all corresponding production logs from the mrt-dashboard UI application.
 - [x]  Cross reference ARKs in the UCSB ETDs collection with those that appear in the Apache access log from the corresponding time period.
 - [x] Attempt reverse `nslookup` of IP addresses for entries in the log that match ARKs in the collection.
- [x] Record any patterns of access that might indicate web scraping.

**Comment by elopatin-uc3** [comment]
Have log files from all three UI servers, from corresponding time frames on my local now. Will run a mySQL query on the UCSB ETDs collection next.

**Comment by elopatin-uc3** [comment]
Per the following MySQL query, I now have a full list of ARKs for objects in the UCSB ETDs collection, saved as a .csv (locally):

`select inv_objects.ark, inv_objects.erc_who, inv_collections.mnemonic
from inv_objects, inv_collections, inv_collections_inv_objects
where inv_objects.id=inv_collections_inv_objects.inv_object_id
and inv_collections.id=inv_collections_inv_objects.inv_collection_id
and inv_collections_inv_objects.inv_collection_id in (47)`

**Comment by elopatin-uc3** [comment]
First script run against logs from `uc3-mrtui01-prd` underway. Running this on `mrtreplic2-dev`.

**Comment by elopatin-uc3** [comment]
All script runs completed. Many dissertation XML files were downloaded between 5-9 times, yet some where not downloaded at all. The IP address associated with downloads (at least in the cases I checked) is an internal one (`172.30.22.60`). I spoke with Scott, and he thinks this may actually be associated with our fixity checking, so I'll follow up with @dloy on Monday when he's back in the office. Note this IP address appears over 400k times in some of the log files.

Another point to mention: if someone was actually crawling the entire collection with the credentials in question, one would think that every single dissertation PDF and XML would have been downloaded at least once. That is not the case, as numerous dissertations were not downloaded at all.

**Comment by elopatin-uc3** [comment]
Attaching script output.
[output.zip](https://github.com/CDLUC3/mrt-doc/files/3799914/output.zip)

**Comment by elopatin-uc3** [comment]
Now searching through access logs from uc3-mrtweb-prd.cdlib.org, as `172.30.22.60` address entries in `mrtuiXX` host logs are the result of a reverse proxy to the load balancer (mrtweb-prd).

**Comment by elopatin-uc3** [comment]
Script run through access logs complete. 

Each of the 3234 ARKs, one per ETD object in the collection, were searched for in the access logs covering the date range of 2019-09-08 through 2019-10-06. These dates overlap the two week period were credentials were posted, beginning on 2019-09-18.

Script output, items of interest:

- 22632 lines total of script output

- 3234 lines dedicated to recording the ARK ending string
- 9702 lines dedicated to recording number of times the ARK was found
- 5244 lines mention "CDLRepositoryHarvester/0.4" (these pertain to IP address 172.30.5.59, which is an internal CDL address for one our eScholarship harvesting processes)
- 1422 lines mention "-" "Ruby" entries (these pertain to IP 128.111.97.63, which maps to UCSB Office of Information Technology, North Hall 2124)
- 3027 blank lines, per ARKs that were not found in the access logs and resulted in 0 count.
- 22629, out of 22632 lines accounted for in above items

The three remaining lines pertain to direct browser access of the collection. These also originate from IPs at the UCSB Office of Information Technology: 128.111.97.50, 128.111.97.63 

`128.111.97.50 - - [29/Aug/2019:13:52:18 -0700] "GET /m/ark%253A%252F13030%252F....... HTTP/1.1" 200 5867 "https://merritt.cdlib.org/s/ucsb_lib_etd?utf8=%E2%9C%93&terms=.......&group=ucsb_lib_etd&commit=Go" "Mozilla/5.0 (Macintosh; Intel Mac OS X 10_13_6) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/76.0.3809.100 Safari/537.36" 0`

`128.111.97.50 - - [29/Aug/2019:16:51:01 -0700] "GET /m/ark%253A%252F13030%252F....... HTTP/1.1" 200 5829 "https://merritt.cdlib.org/collection/search_results?utf8=%E2%9C%93&terms=.......&group=ucsb_lib_etd&commit=Go" "Mozilla/5.0 (Macintosh; Intel Mac OS X 10_13_6) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/76.0.3809.100 Safari/537.36" 0`

`128.111.97.63 - - [11/Sep/2019:12:42:34 -0700] "GET /d/ark%253A%252F13030%252F......./2/producer%252F.......DATA.xml HTTP/1.1" 200 6192 "-" "Mozilla/5.0 (Macintosh; Intel Mac OS X 10_14_6) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/76.0.3809.132 Safari/537.36" 0`

In summary, all 22632 lines of output have been accounted for, and reflect non-malicious content discovery.

**Comment by elopatin-uc3** [comment]
Attaching load balancer script output for reference. ETD info redacted:
[output_web_prod_etd_info_removed.txt.zip](https://github.com/CDLUC3/mrt-doc/files/3822059/output_web_prod_etd_info_removed.txt.zip)
